File: ejemplos_javi/concurrencia_2.cpp
#include <thread>
#include <iostream>
#include <mutex>
#include <memory>

class Number{
    int number;
    std::mutex mtx;

    public:
        explicit Number(int x) : number(x), mtx(){}

        void add(int x){
            std::unique_lock<std::mutex> lock(mtx);
            number += x;
        }

        bool lessthan(int x){
            std::unique_lock<std::mutex> lock(mtx);
            return (number < x);
        }

        void show(){
            std::unique_lock<std::mutex> lock(mtx);
            std::cout << number << std::endl;
        }
};

void add_to_100(std::shared_ptr<Number> num, int n){
    while (num->lessthan(100)){
        num->add(1);
        //std::cout << "Thread " << n << std::endl;
        //num->show();
    }
}

int main(){
    std::shared_ptr<Number> x(new Number(0));
    auto y = x;
    auto z = x;

    std::thread h1(&add_to_100, x, 1);
    std::thread h2(&add_to_100, y, 2);

    h1.join();
    h2.join();

    z->show();

    return 0;
}

File: ejemplos_javi/concurrencia_2.rs
use std::thread;
use std::sync::{Arc, Mutex};

struct Numero{
    value: Mutex<i32>
}

impl Numero{

    fn new(x: i32) -> Numero{
        Numero{value: Mutex::new(x)}
    }

    fn add(& self, x: i32){
        let mut safe_value = self.value.lock().unwrap();
        *safe_value += x;
    }

    fn lessthan(& self, x:i32) -> bool{
        let safe_value = self.value.lock().unwrap();
        return *safe_value < x;
    }

    fn show(& self){
        let safe_value = self.value.lock().unwrap();
        println!("{}", *safe_value);
    }
}

fn add_to_100(x: & Numero, n: i32){
    while x.lessthan(100){
        x.add(1);
        //println!("Thread {}", n);
        //x.show();
    }
}

fn main() {
    let x = Arc::new(Numero::new(0));
    let y = x.clone();
    let z = x.clone();

    let h1 = thread::spawn(move || {
        add_to_100(&x, 1);
    });

    let h2 = thread::spawn(move || {
        add_to_100(&y, 2);
    });

    h1.join().unwrap();
    h2.join().unwrap();

    println!("Final");
    z.show();

}
